### ISSUE TITLE
Mismatch over bridge or tap ifindex

### ISSUE THREAD

**Issue by candlerb** [grnet/snf-nfdhcpd#2]
I'm testing out the synnefo stack in a somewhat unusual config: the ganeti nodes are ubuntu 16.04 (ganeti 2.15.2) with the following sources:

~~~
# for snf-image
deb http://apt.dev.grnet.gr jessie/
deb-src http://apt.dev.grnet.gr jessie/

# for everything else
deb http://apt.dev.grnet.gr snf-dev-0.19/
deb-src http://apt.dev.grnet.gr snf-dev-0.19/
~~~

This has installed the nfdhcpd package version 0.12.4-2~jessie.

Perhaps surprisingly, this all got off to a very good start. Having installed archipelago on the ganeti nodes, they are happily able to launch an image from pithos.

However nfdhcpd has a problem. Firstly it did not start:

~~~
Feb 18 13:54:16 nuc1.ws.nsrc.org nfdhcpd[30672]:  * Starting NFQUEUE-based DHCP/RA server  nfdhcpd
Feb 18 13:54:16 nuc1.ws.nsrc.org nfdhcpd[30672]: Traceback (most recent call last):
Feb 18 13:54:16 nuc1.ws.nsrc.org nfdhcpd[30672]:   File "/usr/sbin/nfdhcpd", line 36, in <module>
Feb 18 13:54:16 nuc1.ws.nsrc.org nfdhcpd[30672]:     import daemon.pidlockfile
Feb 18 13:54:16 nuc1.ws.nsrc.org nfdhcpd[30672]: ImportError: No module named pidlockfile
Feb 18 13:54:16 nuc1.ws.nsrc.org nfdhcpd[30672]:    ...fail!
~~~

I was able to patch this easily enough using the following snippet taken from other synnefo code:

~~~
# Take care of differences between python-daemon versions.
try:
    from daemon import pidfile as pidlockfile
except:
    from daemon import pidlockfile
~~~

and later on change a couple of instances of `daemon.pidlockfile` to `pidlockfile`

Now the problem is that nfdhcpd is not responding. The address was successfully allocated by ganeti:

~~~
root@nuc1:~# gnt-instance add -o snf-image+default --os-parameters                    img_passwd=my_vm_example_passw0rd,img_format=diskdump,img_id=debian_base-7.0-x86_64,img_properties='{"OSFAMILY":"linux"\,"ROOT_PARTITION":"1"}'                    -t plain --disk 0:size=2G --no-name-check --no-ip-check                    --net 0:ip=pool,network=test-net-public          -n nuc1          testvm2
Sat Feb 18 14:08:30 2017  - INFO: NIC/0 inherits netparams ['br255', 'bridged', u'']
Sat Feb 18 14:08:30 2017  - INFO: Chose IP 10.0.0.2 from network test-net-public
Sat Feb 18 14:08:31 2017 * disk 0, size 2.0G
Sat Feb 18 14:08:31 2017 * creating instance disks...
~~~

And the state file created:

~~~
root@nuc1:~# cat /var/lib/nfdhcpd/tap1
INDEV=br255
IP=10.0.0.2
MAC=aa:00:00:e7:68:37
HOSTNAME=testvm2
TAGS=""
GATEWAY=10.0.0.1
SUBNET=10.0.0.0/24
GATEWAY6=
SUBNET6=
EUI64=
~~~

And the state file is being read, and the dhcp packets are being received:

~~~
root@nuc1:~# cat /var/log/nfdhcpd/nfdhcpd.log
...
2017-02-18 14:08:57,103 INFO     Updating configuration for tap1
2017-02-18 14:08:57,104 INFO     Parsing binding file /var/lib/nfdhcpd/tap1
2017-02-18 14:08:57,105 INFO      - Opening L2 socket and binding to tap1
2017-02-18 14:08:57,127 DEBUG     - Getting ifindex for interface tap1 from sysfs
2017-02-18 14:08:57,127 INFO      - Added client 18. hostname testvm2, tap tap1, mac aa:00:00:e7:68:37, ip 10.0.0.2
2017-02-18 14:08:57,127 DEBUG    Pending requests on fds []
2017-02-18 14:09:14,099 DEBUG    Pending requests on fds [5]
2017-02-18 14:09:14,099 INFO      * DHCP: Processing pending request
2017-02-18 14:09:14,104 DEBUG     - Incoming packet from tap with ifindex 16
2017-02-18 14:09:14,104 DEBUG     - Getting binding for ifindex 16
2017-02-18 14:09:14,104 INFO      - No client found. mac: aa:00:00:e7:68:37, ifindex: 16
2017-02-18 14:09:14,105 DEBUG     * Processed 2 requests on NFQUEUE with fd 5
~~~

It looks like the problem is a disagreement over the interface index.  The sys filesystem says the tap is 18 and the bridge is 16:

~~~
root@nuc1:~# cat /sys/class/net/tap1/ifindex
18
root@nuc1:~# cat /sys/class/net/br255/ifindex
16
~~~

The binding has been recorded against the tap (client 18). However nfdhcpd says "incoming packet from tap with ifindex 16", whereas 16 is the bridge not the tap.

**Comment by candlerb** [comment]
Ah I see. Ubuntu has python-nfqueue 0.5-1build2.  The [instructions](https://www.synnefo.org/docs/synnefo/latest/install-guide-debian.html#nfdhcpd) said to use `python-nfqueue=0.4+physindev-1~wheezy`

The wheezy package isn't usable because Ubuntu 16.04 has python 2.7:

~~~
dpkg: dependency problems prevent configuration of python-nfqueue:
 python-nfqueue depends on libpython2.6 (>= 2.6); however:
  Package libpython2.6 is not installed.
~~~

The solution was to use the [trusty package](http://packages.ubuntu.com/trusty/python/python-nfqueue). Now it's happy:

~~~
2017-02-18 15:12:31,463 DEBUG    Pending requests on fds [5]
2017-02-18 15:12:31,463 INFO      * DHCP: Processing pending request
2017-02-18 15:12:31,468 DEBUG    No get_physindev() supported
2017-02-18 15:12:31,468 DEBUG     - Getting binding for mac aa:00:00:e7:68:37
2017-02-18 15:12:31,468 INFO      - Client found. hostname testvm2, tap tap1, mac aa:00:00:e7:68:37, i
p 10.0.0.2
2017-02-18 15:12:31,469 DEBUG     - Getting mac for iface br255
2017-02-18 15:12:31,473 INFO      - DHCP: DHCPDISCOVER from hostname testvm2, tap tap1, mac aa:00:00:e
7:68:37, ip 10.0.0.2
2017-02-18 15:12:31,476 INFO      - RESPONSE: DHCPOFFER for hostname testvm2, tap tap1, mac aa:00:00:e7:68:37, ip 10.0.0.2
~~~

**Comment by kargig** [comment]
Hey Brian, would you mind testing the new (merged) nfdhcpd code/package on Ubuntu?
https://github.com/grnet/nfdhcpd/tree/debian-release-0.16-jessie

**Comment by candlerb** [comment]
I'm not really sure how to apply it.

In the [snf-nfdhcpd](https://github.com/grnet/snf-nfdhcpd/tree/debian) repository, the top-level `nfdhcpd` is a self-contained python module (script), which is simple enough.

However in the [nfdhcpd](https://github.com/grnet/nfdhcpd/tree/debian-release-0.16-jessie) repository you refer to, the top-level `nfdhcpd` is a python package (directory), with the actual code in `nfdhcpd/__init__.py`

So I would have to run it as `python -m nfdhcpd` or `python -m nfdhcpd.__main__` ?

**Comment by candlerb** [comment]
Ah I see:

~~~
    entry_points={
        'console_scripts': ['nfdhcpd = nfdhcpd:main']
    },
~~~

I don't have access right now, but I will try something like:

~~~
pip install git+https://github.com/grnet/nfdhcpd.git@debian-release-0.16-jessie
~~~

**Comment by skalkoto** [comment]
`python setup.py install` or 
`python setup.py develop` should do. We'll add it to pypi to when we have a stable release.

**Comment by candlerb** [comment]
TL;DR: it's not working for me with the nfqueue 0.5, but it does with 0.4.

I updated python-nfqueue back to stock 0.5-1build2 package. Then I updated nfdhcpd as follows:

~~~
mv /usr/sbin/nfdhcpd{,.old}
# to leave the old package installed complete with config files, init scripts etc
pip install git+https://github.com/grnet/nfdhcpd.git@debian-release-0.16-jessie
# that created /usr/local/bin/nfdhcpd
ln -s /usr/local/bin/nfdhcpd /usr/sbin/nfdhcpd
systemctl stop nfdhcpd
systemctl start nfdhcpd
~~~

But the VM doesn't get an IP address.

`/var/log/nfdhcpd/nfdhcpd.log` shows:

~~~
2017-03-31 13:55:48,382 INFO     Starting up nfdhcpd v0.16rc1
2017-03-31 13:55:48,382 INFO     Running as nobody (uid:65534, gid: 65534)
2017-03-31 13:55:48,383 WARNING  dhcp_queue option in ipv6 is deprecated in favor of dhcpv6_queue and may be removed in the future. Please update your configuration
2017-03-31 13:55:48,383 INFO     IPv6 mode set to auto. Detecting suitable mode
2017-03-31 13:55:48,383 INFO     IPv6 detected mode: slaac+dhcpv6
2017-03-31 13:55:48,391 INFO     Setting up NFQUEUE for queue 42, AF 2
2017-03-31 13:55:48,392 DEBUG     - Successfully set up NFQUEUE 42
2017-03-31 13:55:48,392 INFO     Setting up NFQUEUE for queue 43, AF 10
2017-03-31 13:55:48,393 DEBUG     - Successfully set up NFQUEUE 43
2017-03-31 13:55:48,393 INFO     Setting up NFQUEUE for queue 44, AF 10
2017-03-31 13:55:48,393 DEBUG     - Successfully set up NFQUEUE 44
2017-03-31 13:55:48,393 INFO     Setting up NFQUEUE for queue 45, AF 10
2017-03-31 13:55:48,394 DEBUG     - Successfully set up NFQUEUE 45
2017-03-31 13:55:48,394 INFO     Ready to serve requests
2017-03-31 13:55:48,394 DEBUG    Updating configuration for tap1
2017-03-31 13:55:48,395 INFO     Parsing binding file /var/lib/nfdhcpd/tap1
2017-03-31 13:55:48,395 DEBUG     - Opening L2 socket and binding to tap1
2017-03-31 13:55:48,397 WARNING   - Cannot open socket [Errno 19] No such device
2017-03-31 13:55:48,410 DEBUG     - Getting ifindex for interface tap1 from sysfs
2017-03-31 13:55:48,410 ERROR     - tap1 is probably down, removing
2017-03-31 13:55:48,410 WARNING   - Stale configuration for tap1 found
2017-03-31 13:55:48,411 INFO            Key                 Client                  MAC        TAP                   IP                                     IPv6
2017-03-31 13:55:48,411 INFO      * Periodic RA: Starting...
2017-03-31 13:55:48,411 INFO      - RA: Sent 0 RAs in 0.00 seconds
2017-03-31 13:57:10,159 DEBUG    Pending requests on fds []
2017-03-31 13:57:10,169 DEBUG    Updating configuration for tap1
2017-03-31 13:57:10,169 INFO     Parsing binding file /var/lib/nfdhcpd/tap1
2017-03-31 13:57:10,170 DEBUG     - Opening L2 socket and binding to tap1
2017-03-31 13:57:10,186 DEBUG     - Getting ifindex for interface tap1 from sysfs
2017-03-31 13:57:10,186 DEBUG     - Added client 14. hostname testvm2, tap tap1, mac aa:00:00:e7:68:37, ip 10.0.0.2
2017-03-31 13:57:10,187 DEBUG    Pending requests on fds []
2017-03-31 13:57:21,407 DEBUG    Pending requests on fds [5]
2017-03-31 13:57:21,407 INFO      * DHCP: Processing pending request
2017-03-31 13:57:21,411 DEBUG     - Incoming packet from device with ifindex 9
2017-03-31 13:57:21,412 DEBUG     - Binding: Getting binding for ifindex 9
2017-03-31 13:57:21,412 DEBUG     - Binding: No client found for mac:aa:00:00:e7:68:37 / ifindex:9
2017-03-31 13:57:21,412 DEBUG     * Processed 2 requests on NFQUEUE with fd 5
2017-03-31 13:57:29,536 DEBUG    Pending requests on fds [5]
2017-03-31 13:57:29,537 INFO      * DHCP: Processing pending request
2017-03-31 13:57:29,545 DEBUG     - Incoming packet from device with ifindex 9
2017-03-31 13:57:29,545 DEBUG     - Binding: Getting binding for ifindex 9
2017-03-31 13:57:29,546 DEBUG     - Binding: No client found for mac:aa:00:00:e7:68:37 / ifindex:9
2017-03-31 13:57:29,546 DEBUG     * Processed 2 requests on NFQUEUE with fd 5
...
~~~

VM info:

~~~
# gnt-instance info testvm2
...
  NICs:
    - nic/0:
      MAC: aa:00:00:e7:68:37
      IP: 10.0.0.2
      mode: bridged
      link: br255
      vlan:
      network: test-net-public
      UUID: 96dc68a7-c423-4207-b4bb-d0ba9d441c6f
      name: None
~~~

Additional info:

~~~
root@nuc1:~# brctl show br255
bridge name	bridge id		STP enabled	interfaces
br255		8000.c03fd5636411	no		enp3s0.255
							tap1
root@nuc1:~# ls -l /var/lib/nfdhcpd/
total 4
-rw-r--r-- 1 root root 133 Mar 31 13:57 tap1
root@nuc1:~# cat /var/lib/nfdhcpd/tap1
INDEV=br255
IP=10.0.0.2
MAC=aa:00:00:e7:68:37
HOSTNAME=testvm2
TAGS=""
GATEWAY=10.0.0.1
SUBNET=10.0.0.0/24
GATEWAY6=
SUBNET6=
EUI64=
root@nuc1:~# cat /sys/class/net/tap1/ifindex
14
root@nuc1:~# cat /sys/class/net/br255/ifindex
9
~~~

So it looks like as before, nfqueue message has ifindex 9 (= br255), but nfdhcpd expects to associate it with ifindex 14 (tap1)

Finally: I downgraded python-nfqueue package back to `python-nfqueue_0.4-6_amd64.deb` from trusty. And now it's happy:

~~~
2017-03-31 14:11:52,720 INFO     Starting up nfdhcpd v0.16rc1
2017-03-31 14:11:52,721 INFO     Running as nobody (uid:65534, gid: 65534)
2017-03-31 14:11:52,721 WARNING  dhcp_queue option in ipv6 is deprecated in favor of dhcpv6_queue and may be removed in the future. Please update your configuration
2017-03-31 14:11:52,721 INFO     IPv6 mode set to auto. Detecting suitable mode
2017-03-31 14:11:52,721 INFO     IPv6 detected mode: slaac+dhcpv6
2017-03-31 14:11:52,730 INFO     Setting up NFQUEUE for queue 42, AF 2
2017-03-31 14:11:52,730 DEBUG     - Successfully set up NFQUEUE 42
2017-03-31 14:11:52,731 INFO     Setting up NFQUEUE for queue 43, AF 10
2017-03-31 14:11:52,731 DEBUG     - Successfully set up NFQUEUE 43
2017-03-31 14:11:52,731 INFO     Setting up NFQUEUE for queue 44, AF 10
2017-03-31 14:11:52,731 DEBUG     - Successfully set up NFQUEUE 44
2017-03-31 14:11:52,732 INFO     Setting up NFQUEUE for queue 45, AF 10
2017-03-31 14:11:52,732 DEBUG     - Successfully set up NFQUEUE 45
2017-03-31 14:11:52,732 INFO     Ready to serve requests
2017-03-31 14:11:52,733 DEBUG    Updating configuration for tap1
2017-03-31 14:11:52,733 INFO     Parsing binding file /var/lib/nfdhcpd/tap1
2017-03-31 14:11:52,733 DEBUG     - Opening L2 socket and binding to tap1
2017-03-31 14:11:52,742 DEBUG     - Getting ifindex for interface tap1 from sysfs
2017-03-31 14:11:52,742 DEBUG     - Added client aa:00:00:e7:68:37. hostname testvm2, tap tap1, mac aa:00:00:e7:68:37, ip 10.0.0.2
2017-03-31 14:11:52,742 INFO            Key                 Client                  MAC        TAP                   IP                                     IPv6
2017-03-31 14:11:52,742 INFO     aa:00:00:e7:68:37 |              testvm2    aa:00:00:e7:68:37       tap1             10.0.0.2                                     None
2017-03-31 14:11:52,743 INFO      * Periodic RA: Starting...
2017-03-31 14:11:52,743 DEBUG     - RA: Skipping hostname testvm2, tap tap1, mac aa:00:00:e7:68:37, ip 10.0.0.2
2017-03-31 14:11:52,744 INFO      - RA: Sent 0 RAs in 0.00 seconds
2017-03-31 14:11:52,745 DEBUG    Pending requests on fds [5]
2017-03-31 14:11:52,745 INFO      * DHCP: Processing pending request
2017-03-31 14:11:52,749 ERROR    No get_physindev() supported
2017-03-31 14:11:52,749 DEBUG     - Binding: Getting binding for mac aa:00:00:e7:68:37
2017-03-31 14:11:52,749 DEBUG     - Binding: Client found. hostname testvm2, tap tap1, mac aa:00:00:e7:68:37, ip 10.0.0.2
2017-03-31 14:11:52,750 DEBUG     - DHCP: Generating response for hostname testvm2, tap tap1, mac aa:00:00:e7:68:37, ip 10.0.0.2, src 192.0.2.1
2017-03-31 14:11:52,750 DEBUG     - Getting mac for iface br255
2017-03-31 14:11:52,754 INFO      - DHCP: DHCPDISCOVER from hostname testvm2, tap tap1, mac aa:00:00:e7:68:37, ip 10.0.0.2
2017-03-31 14:11:52,756 INFO      - DHCP: DHCPOFFER for hostname testvm2, tap tap1, mac aa:00:00:e7:68:37, ip 10.0.0.2
2017-03-31 14:11:52,760 DEBUG     - Sent 356 bytes on tap1
2017-03-31 14:11:52,760 DEBUG     * Processed 2 requests on NFQUEUE with fd 5
2017-03-31 14:11:52,761 DEBUG    Pending requests on fds [5]
2017-03-31 14:11:52,761 INFO      * DHCP: Processing pending request
2017-03-31 14:11:52,764 ERROR    No get_physindev() supported
2017-03-31 14:11:52,764 DEBUG     - Binding: Getting binding for mac aa:00:00:e7:68:37
2017-03-31 14:11:52,764 DEBUG     - Binding: Client found. hostname testvm2, tap tap1, mac aa:00:00:e7:68:37, ip 10.0.0.2
2017-03-31 14:11:52,764 DEBUG     - DHCP: Generating response for hostname testvm2, tap tap1, mac aa:00:00:e7:68:37, ip 10.0.0.2, src 192.0.2.1
2017-03-31 14:11:52,764 DEBUG     - Getting mac for iface br255
2017-03-31 14:11:52,768 INFO      - DHCP: DHCPREQUEST from hostname testvm2, tap tap1, mac aa:00:00:e7:68:37, ip 10.0.0.2
2017-03-31 14:11:52,770 INFO      - DHCP: DHCPACK for hostname testvm2, tap tap1, mac aa:00:00:e7:68:37, ip 10.0.0.2
2017-03-31 14:11:52,774 DEBUG     - Sent 356 bytes on tap1
2017-03-31 14:11:52,774 DEBUG     * Processed 2 requests on NFQUEUE with fd 5
~~~

Notice how in this case it says `Getting binding for mac ...` rather than `Getting binding for ifindex 9`

**Comment by kargig** [comment]
Brian: can we continue this discussion on the proper repo? https://github.com/grnet/nfdhcpd.
Would you be willing to create a new issue there please and also move the relevant parts of your report there ?

Afaik only upstream python-nfqueue 0.6+ has the appropriate patch for physindev:
https://github.com/chifflier/nfqueue-bindings/commit/fae71ad82243d78895aed5b3d038425c942af392

so one needs either a patched package for any Debian that is not newer than Stretch or Ubuntu that is not newer than yakety (16.10).

Open an issue at the proper repo and we wil correct the documentation accordingly. Thanks!
